crates/storage_impl/src/errors.rs

```diff
@@ -53,7 +58,15 @@ impl From<diesel::result::Error> for StorageError {
 
 impl From<error_stack::Report<DatabaseError>> for StorageError {
     fn from(err: error_stack::Report<DatabaseError>) -> Self {
-        Self::DatabaseError(err)
+        match err.current_context() {
+            DatabaseError::DatabaseConnectionError => Self::DatabaseConnectionError,
+            DatabaseError::NotFound => Self::ValueNotFound(String::from("db value not found")),
+            DatabaseError::UniqueViolation => Self::DuplicateValue {
+                entity: "db entity",
+                key: None,
+            },
+            err => Self::DatabaseError(error_stack::report!(*err)),
```

Comment:
We'd lose the original error context by doing this.

```suggestion
            _ => Self::DatabaseError(err),
```